crates/ruff/src/checkers/ast.rs

```diff
@@ -136,163 +104,66 @@ impl<'a> Checker<'a> {
             noqa,
             path,
             package,
-            module_path,
+            module_path: module_path.clone(),
             is_interface_definition,
             locator,
             stylist: style,
             indexer,
+            ctx: AstContext {
```

Comment:
Could we move the `context` out of the `Checker` entirely and pass it around instead of the checker so that:

* The checker orchestrates the "checking"
* `context` stores the state and is passed to the rules.

Reply:
In that setup, would `ctx` be passed around as part of the `Visitor` trait?

Reply:
I should have spent some more time reading the code. Storing the `context` on the `Visitor` is probably fine. What I'm mainly wondering is if it would be possible to pass the `context` to various rules instead of the `Checker`.

Reply:
👍 That's my hope, that we can stop passing `Checker`, and instead pass: `AstContext`, a mutable `diagnostics` vector, and some kind of settings representation.

(Beyond that, we'd also need `Locator`, `Stylist,` etc., which currently live on `Checker` but there's no reason they need to.)